tasmota/tasmota_xsns_sensor/xsns_115_wooliis.ino

```diff
@@ -0,0 +1,230 @@
+
+  Wooliis.charge_percent = buffer[4];
+  Wooliis.remaining_capacity = 0.1f * ((uint16_t)buffer[5]<<8 | buffer[6]);
+  Wooliis.voltage = 0.1f * ((uint16_t)buffer[7]<<8 | buffer[8]);
+  Wooliis.current = (buffer[19] ? -0.1f : 0.1f) * ((uint16_t)buffer[9]<<8 | buffer[10]);
+  Wooliis.energy_in = 0.1f * ((uint32_t)buffer[11]<<16 | (uint16_t)buffer[12]<<8 | buffer[13]);
+  Wooliis.energy_out = 0.1f * ((uint32_t)buffer[14]<<16 | (uint16_t)buffer[15]<<8 | buffer[16]);
+  Wooliis.status = buffer[19];
+
+  Wooliis.valid = 1;
+
+  return;
+}
+
+/*********************************************************************************************/
+
+void WooliisSecond(void)                 // Every second
+{
+  if (Wooliis.ready)
+    WooliisReadData();
+}
+
+/*********************************************************************************************/
+
+void WooliisInit(void) {
+  if (PinUsed(GPIO_WOOLIIS_RX)) {
+    WooliisSerial = new TasmotaSerial(Pin(GPIO_WOOLIIS_RX), -1, 1);
+    if (WooliisSerial->begin(9600)) {
+      if (WooliisSerial->hardwareSerial()) { ClaimSerial(); }
+      WooliisSerial->flush();  // Clear serial buffer
+      Wooliis.ready = 1;
+      AddLog(LOG_LEVEL_DEBUG, PSTR("Wooliis: Serial ready"));
+    }
+  }
+  Wooliis.valid = 0;
+}
+
+#ifdef USE_WEBSERVER
+const char HTTP_SNS_WOOLIIS_BCP[] = "{s}%s " D_BATTERY_CHARGE "{m}%d " D_UNIT_PERCENT "{e}";
+const char HTTP_SNS_WOOLIIS_CAP[] = "{s}%s " D_CAPACITY "{m}%s " D_UNIT_CHARGE "{e}";
+const char HTTP_SNS_WOOLIIS_CHG[] = "{s}%s " D_CHARGING "{m}%s{e}";
+const char HTTP_SNS_WOOLIIS_VOL[] = "{s}%s " D_VOLTAGE "{m}%s " D_UNIT_VOLT "{e}";
+const char HTTP_SNS_WOOLIIS_CUR[] = "{s}%s " D_CURRENT "{m}%s " D_UNIT_AMPERE "{e}";
+const char HTTP_SNS_WOOLIIS_POW[] = "{s}%s " D_POWERUSAGE "{m}%s " D_UNIT_WATT "{e}";
+const char HTTP_SNS_WOOLIIS_EIN[] = "{s}%s " D_ENERGY " " D_INGRESS "{m}%s " D_UNIT_WATTHOUR "{e}";
+const char HTTP_SNS_WOOLIIS_EEG[] = "{s}%s " D_ENERGY " " D_EGRESS "{m}%s " D_UNIT_WATTHOUR "{e}";
+#endif  // USE_WEBSERVER
+
+void WooliisShow(bool json) {
+  char float_str[8];
+  if (Wooliis.valid) {
+    if (json) {
+      ResponseAppend_P(PSTR(",\"%s\":{"), D_WOOLIIS);
+      dtostrfd(Wooliis.remaining_capacity, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_CAPACITY, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%d,"), D_CHARGING, (Wooliis.status ? 1 : 0));
+      dtostrfd(Wooliis.voltage, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_VOLTAGE, float_str);
+      dtostrfd(Wooliis.current, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_CURRENT, float_str);
+      dtostrfd(Wooliis.voltage*Wooliis.current, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_POWERUSAGE, float_str);
+      dtostrfd(Wooliis.energy_out, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s"), D_ENERGY, float_str);
+      dtostrfd(Wooliis.energy_in, 1, float_str);
+      ResponseAppend_P(PSTR("\"%s\":%s,"), D_ENERGY "_returned", float_str);
```

Comment:
Do not make composite fieldnames. To keep using already known names and keeping code as small as possible I would suggest to use ``D_JSON_EXPORT`` here.

Reply:
Alright, so for energy going into the battery I will use _IMPORT and for energy supplied by the battery I will use _EXPORT. Does that make sense?

Reply:
Yes.

Reply:
@arendst this is the last question for you, all the rest has been modified and tested.

Reply:
yes

Reply:
Oh well, so if yes means sensor code, it's done, here are the definitions:

#define D_IMPORT "Import"
#define D_EXPORT "Export"
#define D_JSON_CAPACITY "Capacity"
#define D_JSON_CHARGING "Charging"

If that was the wrong interpretation, I will change it.
Anything else to be modified?

Reply:
You've done enough to make it work.

I'll merge this and will change some duplicates to lower the codebase.

Thx.